### ISSUE TITLE
bug(modal): the first press of shift + tab does not focus an element in the modal, focus is not trapped

### ISSUE THREAD

**Issue by maxkorz** [ng-bootstrap/ng-bootstrap#2718]
https://stackblitz.com/angular/ygpjvdbxbre

1. Launch demo modal
2. Press <kbd>Tab</kbd> - first element in the modal container gets the focus - focus is trapped ✔
3. Keep pressing <kbd>Tab</kbd> - focus is looped ✔
4. Press <kbd>Shift</kbd> + <kbd>Tab</kbd> - focus is going to a previous element, focus is still trapped  ✔

<hr>

1. Launch demo modal
2. Press <kbd>Shift</kbd> + <kbd>Tab</kbd> - element outside of the modal container gets the focus - focus is not trapped ❌

P.S. it works fine in [Datepicker in a popup](https://stackblitz.com/angular/nvpvpagxmoxo). 
1. Click on the calendar icon
2. Press <kbd>Shift</kbd> + <kbd>Tab</kbd> - focus is trapped and looped ✔

**Comment by maxkorz** [comment]
@benouat I don't understand. How come <kbd>Tab</kbd> is "_a side-effect, not a feature_", if focus trapping is intentional? (`ngbFocusTrap` is used for this)

<hr>

by the way, when you do `confirm()`, the focus is completely trapped inside the native modal dialog.

**Comment by benouat** [comment]
@maxokorokov my point is that from an API stand point, user should always specify/manage themselves what to be focused when modal is opened. I mean it should be mandatory somehow. They should never open a modal without handling that.

The fact that if you do open a modal without specifying an element to be focused is working when first pressing <kbd>Tab</kbd> **is** the side effect to me

**Comment by maxkorz** [comment]
@benouat 

> Maybe we should implement a kind of NgbAutofocus as a directive hosted on [autofocus] that will perform the focus imperatively on next tick.

I don't think that this is a good idea. According to https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-autofocus
> Only one element in a document can have the autofocus attribute. It cannot be applied if the type attribute is hidden, because hidden inputs cannot be focused. 

> Automatically focusing a form control can confuse visually-impaired people who using screen-reading technology. When autofocus is assigned, screen-readers "teleport" their user to the form control without warning them beforehand.

**Comment by benouat** [comment]
@maxkorz visually-impaired or not, we do want our user to be "teleported" into the content of the modal.

I still think the directive is a good idea, BTW material implemented it for their accessible focus management

**Comment by maxkorz** [comment]
@benouat I just checked, ngbFocusTrap works fine in the datepicker:

https://stackblitz.com/angular/nvpvpagxmoxo
1. Click on the calendar icon
2. Press Shift + Tab - focus is trapped and looped ✔

(I updated my first message)

**Comment by benouat** [comment]
yes it is working in the datepicker because of what I described: 

> my point is that from an API stand point, user should always specify/manage themselves what to be focused when modal is opened

In this case, it is not the user, but the datepicker itself that will focus the selected date when opened. Something is then initially focused, `ngbFocusTrap` can work properly.

Again, issue for modal is that if you don't focus something when it is opened, `ngbFocusTrap` won't work properly.

**Comment by maxkorz** [comment]
@benouat I still think that the issue is in the `ngbFocusTrap`.
`ngbFocusTrap` is described as `Function that enforces browser focus to be trapped inside a DOM element.`. Which does not happen for modals.

Small fix in `ngbFocusTrap` fixes the issue
```
....
      .subscribe(([tabEvent, focusedElement]) => {
        const[first, last] = getFocusableBoundaryElements(element);

        // trap focus on the first press of Shift + Tab
        if (element === tabEvent.target && tabEvent.shiftKey) {
          last.focus();
          tabEvent.preventDefault();
        }
...
```

What do you think?

**Comment by maxkorz** [comment]
> I know that handling the other one with just Tab is already working (because this is the default browser behaviour), but for consistency, if we implement one, we have to implement the other one that will just override the default behaviour

I agree 👍

<hr>

> when using modal, user should take care (in a kind of mandatory way) of focusing something inside the modal.

I strongly disagree 😃

**Comment by pkozlowski-opensource** [comment]
> Nonetheless, documentation should clearly state that when using modal, user should take care (in a kind of mandatory way) of focusing something inside the modal.

We could strongly suggest this but IMO modal should have sensible defaults (ex. first focusable element or modal window itself (as today).

**Comment by maxkorz** [comment]
@benouat can you open a PR with your solution?

**Comment by benouat** [comment]
@maxkorz yes sure, i will !

**Comment by mendeza** [comment]
Hi Folks,
My team at Thomson Reuters has just retained an Accessibility testing expert who spent years working on Government projects. I asked him to look at this issue. His feedback:

Once inside the Modal, focus _must_ be trapped; SHIFT-TAB immediately after entry would give focus to the last focusable item at the bottom/end of the Modal (wrap-around).

Once the Modal is spawned, focus _must_ go to the _first_ actionable item - presumably the close (X) button. FYI, arbitrarily focusing some element in the middle of content could be troublesome to exit if the Modal was entered by error.

A suggestion he made is to set the `aria-describedby` attribute on the close button and reference the ID of the modal title. I realize that's a problematic thing to enforce, though you could show examples in your demos. What makes this a bit better than the `aria-label="Close"` that you currently use, is it provides more context to users, so that they can be exactly sure of what they are closing when focus lands on the close button.

I hope this is helpful.

**Comment by CBly249013** [comment]
Something like

<div><h1 id="M1">Modal title</h1>
<span ARIA-label="close" ARIA-describedby="M1"> x</span>
</div>

Or if ARIA-describedby can't be used

<div><h1>Modal title</h1>
<span ARIA-label="close modal title"> x</span>
</div>

**Comment by benouat** [comment]
Thanks for you input @mendeza ! Much appreciated here

Now after a few days step back, let me try to explain why I am reluctant to perform the change I described and we discussed previously in this issue using your comments: 

> Once inside the Modal, focus _must_ be trapped;

We are fine here, `NgbFocusTrap` is doing that today. Important words here are "**Once inside**".


> Once the Modal is spawned, focus _must_ go to the first actionable item

This is what we thought would be enough by focusing the modal window (@pkozlowski-opensource we don't do first focusable). It is apparently not enough and should be changed (disclaimer, it is **not** a framework only task, ng-bootstrap user must also be involved)

1. This is why I have been repeating myself from the beginning of this issue, as a framework, we can't (or we could, but don't) know what the modal content contains, so to make sure focus is going somewhere (important words here "**focus must go**") users should take responsibility to focus themselves something (ie _going inside_).
Then `NgbFocusTrap` will do its job, <kbd>Shift</kbd>+<kbd>Tab</kbd> will work as expected.

2. To help users doing that properly, this is why I suggested to introduce an `NgbAutofocus` directive that will do that: simply set the focus to the first actionable item, defined by the user using the directive, as soon as the modal is opened/shown

**Comment by maxkorz** [comment]
@benouat Lets say you add the `NgbAutofocus` directive. But you can't force a user to use it (I won't use it for sure). And the bug will not be fixed. Once the focus is trapped, it shouldn't be possible to focus an element outside of the trapped element. Once the modal is opened, you shouldn't be able to focus an element outside of the modal window. But you can do that with <kbd>Shift</kbd> + <kbd>Tab</kbd>. And that is a bug that can be fixed with 10 lines of code.

You saw the bug in the demo. You can focus an element "behind" the modal window. You can't blame a user for that.

**Comment by pkozlowski-opensource** [comment]
IMO we should do the following:
* add a `NgbAutofocus` directive so people can indicate an element inside modal to be auto-focused;
* if the `NgbAutofocus` directive is not used we should focus first focusable element.

"we should focus first focusable element" will be painful to do, though, as it will require lots of low-level DOM querying.

**Comment by maxokorokov** [comment]
As everybody is commenting this, I'll leave my 2 cents too :)

- add the `NgbAutoFocus` directive
- optionally focus first focusable element, if user desires to do so. I wouldn't try focusing first element by default, because I have absolutely no idea what the modal content might be.

Something like:
```ts
modalService.open(content, {focusFirst: true})
```

This issue is not a bug to me, because ONCE anything is focused in the popup, focus is trapped. There is nothing focused initially. 

I prefer this to be closed as `wontfix` and open a separate feature request with clear desired API.

**Comment by maxokorokov** [comment]
Opened #2728

**Comment by peterblazejewicz** [comment]
Hi folks,
I'm under impression that SHIFT + TAB allows me to quickly start from the bottom most interactive control and ease navigation to the option of my choice, simply by going backward (I hit the MacOS about this mac and start backward navigation with no no fuss).
the problem is that implementation of the focus trap seems to expect users to start with `Tab`, while disregarding `Tab + SHIFT` option.
I'd like to see a code, like one discussed above being, added:
https://github.com/ng-bootstrap/ng-bootstrap/issues/2718#issuecomment-422322788
thanks!

ref: 
https://github.com/twbs/bootstrap/issues/25277#issuecomment-357097404
https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal

**Comment by benouat** [comment]
@peterblazejewicz here are a few highlight

> the problem is that implementation of the focus trap seems to expect users to start with Tab, while disregarding Tab + SHIFT option

This is a wrong assumption. Our focus-trap implementation works great with both <kbd>Tab</kbd> and <kbd>Shift</kbd>+<kbd>Tab</kbd> **once and only once** the focus is locted anywhere inside (not inclusive) the element that host the focus trap.
